src/e2e/resources/simple-temple-expected/simple-temple-test-user/dao/dao.go

```diff
@@ -188,6 +191,13 @@ func (dao *DAO) CreateSimpleTempleTestUser(input CreateSimpleTempleTestUserInput
 	var simpleTempleTestUser SimpleTempleTestUser
 	err := row.Scan(&simpleTempleTestUser.ID, &simpleTempleTestUser.SimpleTempleTestUser, &simpleTempleTestUser.Email, &simpleTempleTestUser.FirstName, &simpleTempleTestUser.LastName, &simpleTempleTestUser.CreatedAt, &simpleTempleTestUser.NumberOfDogs, &simpleTempleTestUser.Yeets, &simpleTempleTestUser.CurrentBankBalance, &simpleTempleTestUser.BirthDate, &simpleTempleTestUser.BreakfastTime)
 	if err != nil {
+		// pq specific error
+		if err, ok := err.(*pq.Error); ok {
+			if err.Code.Name() == psqlUniqueViolation {
+				return nil, ErrDuplicateSimpleTempleTestUser("")
```

Comment:
Why the empty string?

Reply:
In the request handler, we match against the type of error and return a response accordingly:
```go
if err != nil {
	switch err.(type) {
	 case dao.ErrSimpleTempleTestUserNotFound:
	         // Return a 404 
         case dao.ErrDuplicateSimpleTempleTestUser:
	         // Return a 403
         default:
                  // Return a 500
	}
}
```

To do that, I define a new type in `errors.go`: 

```go
type ErrDuplicateSimpleTempleTestUser string
```

This requires an existing type for it to be defined against, and string made the most sense to allow for some additional data in the future (such as where the offending constraint was violated etc.)

The alternative to this is to use an error builder:

```go
return nil, errors.New("Error description")
```

The issue with this is that you can no longer pattern match on the type (since it's always Error), and therefore in the handler we can't distinguish between an internal server error (i.e. a 500) and a duplicate user (i.e a 403). We could avoid this by wrapping the error with an associated code of some description, then matching on this code and returning appropriately, but that felt like a bit of over engineering for this small use case, and would increase the complexity of the handler.

Reply:
Aaaand, scratch all of that. `pq` exposes a "Detail" of the error which I didn't see before.

The error is now:

```
Key (bar)=(123) already exists.
```